src/parquet/arrow/reader.cc

```diff
@@ -521,93 +534,182 @@ Status FlatColumnReader::Impl::TypedReadBatch(
+        (descr_->max_repetition_level() == 1)) {
+      can_parse = true;
+      minimal_valid_def_level--;
+      nullable_lists = false;
+    } else if ((descr_->max_definition_level() == 2) &&
+               (descr_->schema_node()->is_required()) &&
+               (descr_->max_repetition_level() == 1)) {
+      can_parse = true;
+    } else if ((descr_->max_definition_level() == 3) &&
+               (descr_->max_repetition_level() == 1)) {
+      can_parse = true;
+    } else if ((descr_->max_definition_level() == 1) &&
+               (descr_->max_repetition_level() == 1)) {
+      can_parse = true;
+      minimal_valid_def_level--;
+      nullable_lists = false;
     }
+
+    if (can_parse) {
+      std::shared_ptr<Array> values_array = *array;
+      int32_t offset = 0;
+      ::arrow::Int32Builder offset_builder(pool_, ::arrow::int32());
+      for (int64_t i = 0; i < total_levels_read; i++) {
+        if ((rep_levels[i] == 0) || (i == 0)) {
+          // New List
+          offset_builder.Append(offset);
+        }
+        if (def_levels[i] >= minimal_valid_def_level) {
+          // We have a (possibly null) element on the lowest nesting level
+          offset++;
+        }
+      }
+      offset_builder.Append(offset);
+      std::shared_ptr<Array> offset_array;
+      offset_builder.Finish(&offset_array);
+      std::shared_ptr<Buffer> offset_buffer =
+          static_cast<Int32Array*>(offset_array.get())->data();
+
+      // Build list null bitmap, if def_levels == 0, then the list is null.
+      int valid_lists_size = ::arrow::BitUtil::CeilByte(offset_array->length()) / 8;
+      auto valid_lists = std::make_shared<PoolBuffer>(pool_);
+      int null_lists_count = 0;
+      if (nullable_lists) {
+        RETURN_NOT_OK(valid_lists->Resize(valid_lists_size));
+        uint8_t* valid_lists_ptr = valid_lists->mutable_data();
+        memset(valid_lists_ptr, 0, valid_lists_size);
+        int valid_lists_idx = 0;
+        for (int64_t i = 0; i < total_levels_read; i++) {
+          if ((rep_levels[i] == 0) || (i == 0)) {
+            if (def_levels[i] > 0) {
+              ::arrow::BitUtil::SetBit(valid_lists_ptr, valid_lists_idx);
+            } else {
+              null_lists_count++;
+            }
+            valid_lists_idx++;
+          }
+        }
+      }
+
+      *array = std::make_shared<ListArray>(::arrow::list(field_->type),
+          offset_array->length() - 1, offset_buffer, values_array, null_lists_count,
+          valid_lists);
+    } else {
+      return Status::NotImplemented(
+          "Only flat lists with a max definition level of 1, 2 or 3 are supported yet");
+    }
+  }
+  return Status::OK();
+}
+
+template <typename ArrowType, typename ParquetType>
+Status ColumnReader::Impl::TypedReadBatch(int batch_size, std::shared_ptr<Array>* out) {
+  using ArrowCType = typename ArrowType::c_type;
+
+  int values_to_read = batch_size;
```

Comment:
Until batch_size always refers to the number of logical rows, it will be a source of issues. Doesn't have to get done in this patch but it will need to maintain state in between batch read calls